### PR TITLE
[Multiple DataSource] Integrate multiple datasource with dev tool console

### LINKED ISSUE #2699: [MD]Multi data source integration with Dev tool

### PR DESCRIPTION
**Pull request by zhongnansu**
### Description
[Multiple DataSource] Integrate multiple datasource with dev tool console

### Update:
After #3775 fixes autocomplete issue, I updated this PR to adopt the change, autocomplete also works for datasources now

Video

https://user-images.githubusercontent.com/32652829/229190663-036b6e68-8a04-4d3c-9dbd-e0b0cf45280e.mov

### Issues Resolved
#2699
part of #3577 
 
### Check List
- [ ] All tests pass
  - [ ] `yarn test:jest`
  - [ ] `yarn test:jest_integration`
  - [ ] `yarn test:ftr`
- [ ] New functionality includes testing.
- [ ] New functionality has been documented.
- [ ] Update [CHANGELOG.md](./../CHANGELOG.md)
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Comment by zhongnansu** [comment]
cc: @KrooshalUX for UI/UX feedback

File: src/plugins/console/server/routes/api/console/proxy/create_handler.ts [hunk 1]
```diff
@@ -130,7 +136,7 @@ export const createHandler = ({
     });
   } catch (e: any) {
     const isResponseErrorFlag = isResponseError(e);
-
+    if (!isResponseError) log.error(e);
```
Comment (kristenTian) [inline src/plugins/console/server/routes/api/console/proxy/create_handler.ts:139]: here looks like - if not error, log error?

File: src/plugins/dev_tools/public/application.tsx [hunk 2]
```diff
@@ -63,25 +99,100 @@ function DevToolsWrapper({ devTools, activeDevTool, updateRoute }: DevToolsWrapp
...
+                    {currentDevTool.title}
+                  </EuiTab>
+                </EuiToolTip>
+              ))}
+            </EuiTabs>
+          </EuiFlexItem>
+          {dataSourceEnabled ? (
+            <EuiFlexItem grow={false} className="dataSourceSelector">
+              <EuiComboBox
+                aria-label="Select a Data Source"
```
Comment (kristenTian) [inline src/plugins/dev_tools/public/application.tsx:181]: nit, translation support

File: src/plugins/console/server/routes/api/console/proxy/tests/query_string.test.ts [hunk 3]
```diff
@@ -63,7 +64,10 @@ describe('Console Proxy Route', () => {
       describe('contains full url', () => {
         it('treats the url as a path', async () => {
           await request('GET', 'http://evil.com/test');
+          // const res = opensearchClient.asCurrentUser.transport.request.mock.calls;
```
Comment (kristenTian) [inline src/plugins/console/server/routes/api/console/proxy/tests/query_string.test.ts:67]: nit

File: src/plugins/console/server/routes/api/console/proxy/tests/query_string.test.ts [hunk 4]
```diff
@@ -42,13 +42,14 @@ describe('Console Proxy Route', () => {
   beforeEach(() => {
     const requestHandlerContextMock = coreMock.createRequestHandlerContext();
     opensearchClient = requestHandlerContextMock.opensearch.client;
+    // buildBufferedBodyMock.mockResolvedValue(Buffer.from('body'));
```
Comment (kristenTian) [inline src/plugins/console/server/routes/api/console/proxy/tests/query_string.test.ts:45]: ditto

File: src/plugins/console/server/routes/api/console/proxy/create_handler.ts (hunk 1 above)
Comment (zhongnansu) [reply src/plugins/console/server/routes/api/console/proxy/create_handler.ts:139]: if response error, we return to client side, don't log
If it's not response error, we add a log here for troubleshoot

**Comment by KrooshalUX** [comment]
LGTM from UX perspective - small nit, but it looks like the use of "tabs" means that the horizontal rule is getting truncated in order to accommodate the picker. Not blocking now, but if its an easy fix to remove the Tab, change "console" into a header, and place in a horizontal rule, that would be a nice touch.

File: src/plugins/dev_tools/public/application.tsx (hunk 2 above)
Comment (zhongnansu) [reply src/plugins/dev_tools/public/application.tsx:181]: added

**Review by kristenTian — APPROVED** [review]

**Review by kristenTian — APPROVED** [review]

**Comment by zhongnansu** [comment]
> LGTM from UX perspective - small nit, but it looks like the use of "tabs" means that the horizontal rule is getting truncated in order to accommodate the picker. Not blocking now, but if its an easy fix to remove the Tab, change "console" into a header, and place in a horizontal rule, that would be a nice touch.

The tricky part is, tab is generated not by `console` plugin, but by `dev_tool` plugin. There's a chance that other plugins are relying `dev_tool` plugin to register themselves, which will generate another tab in dev tool page. If we change this, it will break the contract with existing plugins. I'd suggest we review the blast radius and have the change in `3.0.0` @KrooshalUX

**Review by kristenTian — APPROVED** [review]

File: src/plugins/dev_tools/public/application.tsx [hunk 5]
```diff
@@ -51,8 +73,22 @@ interface MountedDevToolDescriptor {
...
+  devTools,
+  activeDevTool,
+  updateRoute,
+  savedObjects,
+  notifications: { toasts },
+  dataSourceEnabled,
+}: DevToolsWrapperProps) {
   const mountedTool = useRef<MountedDevToolDescriptor | null>(null);
+  const [dataSources, setDataSources] = useState<DataSourceOption[]>([]);
+  const [selectedOptions, setSelected] = useState<DataSourceOption[]>([]);
```
Comment (abbyhu2000) [inline src/plugins/dev_tools/public/application.tsx:91]: nit: `setSelectedOptions`?

File: src/plugins/console/public/plugin.ts [hunk 6]
```diff
@@ -62,7 +62,7 @@ export class ConsoleUIPlugin implements Plugin<void, void, AppSetupUIPluginDepen
         defaultMessage: 'Console',
       }),
       enableRouting: false,
-      mount: async ({ element }) => {
+      mount: async ({ element, dataSourceId }) => {
         const [core] = await getStartServices();
```
Comment (abbyhu2000) [inline src/plugins/console/public/plugin.ts:67]: Should we follow the format as other plugins here? 
```
mount: async (params: AppMountParameters) => {
        const { element, dataSourceId } = params;
```

**Comment by KrooshalUX** [comment]
> > LGTM from UX perspective - small nit, but it looks like the use of "tabs" means that the horizontal rule is getting truncated in order to accommodate the picker. Not blocking now, but if its an easy fix to remove the Tab, change "console" into a header, and place in a horizontal rule, that would be a nice touch.
> 
> The tricky part is, tab is generated not by `console` plugin, but by `dev_tool` plugin. There's a chance that other plugins are relying `dev_tool` plugin to register themselves, which will generate another tab in dev tool page. If we change this, it will break the contract with existing plugins. I'd suggest we review the blast radius and have the change in `3.0.0` @KrooshalUX

Gotcha, that is helpful context.  I have some ideas that I can take to the feature owners for further exploration.

File: src/plugins/dev_tools/public/application.tsx (hunk 5 above)
Comment (zhongnansu) [reply src/plugins/dev_tools/public/application.tsx:91]: good catch, let me fix this

**Review by abbyhu2000 — DISMISSED** [review]

File: src/plugins/console/public/plugin.ts (hunk 6 above)
Comment (zhongnansu) [reply src/plugins/console/public/plugin.ts:67]: I'd rather keep it, reducing 1 more line of code~

**Review by kristenTian — APPROVED** [review]

**Review by abbyhu2000 — APPROVED** [review]

**Comment by zhongnansu** [comment]
address the comment, CI should be good, since the change is only function naming, merging

**Pull request merged**